plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.rules.model;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchType;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Simple abstract class that represent record stored in the Query Insight Framework
+ *
+ * @param <T> The value type associated with the record
+ * @opensearch.internal
+ */
+public abstract class SearchQueryRecord<T extends Number & Comparable<T>>
```

Comment:
I'm not convinced that the `T` parameter on the base class is a good idea. 

If a subclass wants to add fields, isn't that their responsibility?

Reply:
IMHO, similar to metrics, all records should have:
- a value that this record holds.
- attributes associated with this record

This `SearchQueryRecord` class should define those common attributes and also the generic value `T`. 

Also, the `compareTo` and `equals` function would be almost the same for most of the subclasses. Thus I want to do it in the parent class level to avoid duplication.

Reply:
> a value that this record holds.

Why should a record be limited to a single value?

Reply:
Make sense! I just refactored the code based on your and rishabhmaurya's suggestions to make the records more generic!